Improve State Machine Controller Safety, Reduce Implicit Behaviour

olivia-banks opened the pull request:
Title. Before, the first and last states we just guessed at, now they have to explicitly defined. States can now be colored in the generated Graphviz DOT graph. Duplicate states return an error instead of silently failing. Duplicate state roles return an error instead of silently being overwritten.

Commit:
[smc] base graphviz DOT generation

Commit:
[smc] improve safety, reduce implicit behaviour

Commit:
[git] rebase master

Commit:
Merge branch 'master' into milo-sm-ano

Commit:
[smc] add graphviz DOT styling

teowelton commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/fsm/InsaneStateMachineException.java:
```diff
@@ -0,0 +1,10 @@
+ */
+public class InsaneStateMachineException extends RuntimeException {
+    public InsaneStateMachineException(String reason) {
+        super("Encountered an insane state machine: " + reason + ".");
```
This is for my own knowledge- but what does super mean?

olivia-banks replied:
Totally! Since this class extends `RuntimeException`, the `super` in this constructor calls the matching constructor in the `RuntimeException` abstract class. It's sorta like:

```java
this = new RuntimeException("Encountered an insane state machine: " + reason + ".");
```

But actually valid.

Commit:
[smc] state usage -> state role

Commit:
[smc] update smc example to conform to updated api

olivia-banks commented:
Code is working without modification.

Commit:
[smc] simplify

Commit:
[smc] clarify exception type names

Commit:
[smc] eliminate excess exception type

blazeboy75 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/fsm/StateMachine.java:
```diff
@@ -1,26 +1,102 @@
+
     /* The flat list of states. The key is the state, the value is if it's been reached. */
-    private final HashMap<State, Boolean> states_reached = new HashMap<>();
+    private final HashMap<State, Boolean> statesReached = new HashMap<>();
```
we want the start methods to run again- (remember the meeting milo)

olivia-banks replied:
Fixed in [`b177523`](https://github.com/rh-robotics/19922-CSRC/pull/42/commits/b17752345df9febcdbc5c42d3553eb8f946ae00f).

blazeboy75 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/fsm/StateMachine.java:
```diff
@@ -32,42 +108,50 @@ public StateMachine addState(State state) {
      * @return The state machine.
      */
     public StateMachine deleteState(State state) {
```
wwhat- why? Just delete the class that declares the state

olivia-banks replied:
Fixed in [`b177523`](https://github.com/rh-robotics/19922-CSRC/pull/42/commits/b17752345df9febcdbc5c42d3553eb8f946ae00f).

Commit:
[smc] address reviews

init() now takes the place of start(), with start() being run on each
state change.
